File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.dart
class Solution {
 List<int> topKFrequent(List<int> nums, int k) {
  Map<int, int> freq = {};
  for (int num in nums) {
    freq[num] = freq.containsKey(num) ? freq[num]! + 1 : 1;
  }

  List<List<int>> buckets = List.filled(nums.length + 1, []);

  for (int key in freq.keys) {
    int frequency = freq[key]!;
    if (buckets[frequency].isEmpty) {
      buckets[frequency] = [key];
    } else {
      buckets[frequency].add(key);
    }
  }

  List<int> result = [];
  for (int i = buckets.length - 1; i >= 0 && result.length < k; i--) {
    if (buckets[i].isNotEmpty) {
      result.addAll(buckets[i]);
    }
  }

  return result;
}

}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
         unordered_map<int,int> frequncyMap;

         for(int num : nums){
            frequncyMap[num]++;
         }


         priority_queue<pair<int,int>,vector<pair<int, int>> ,greater <pair<int,int>>> minHeap;

         for(const auto& entry : frequncyMap){
            minHeap.push({entry.second,entry.first});
            if(minHeap.size()>k){
                minHeap.pop();
            }
         }

         vector<int> result;

         while(!minHeap.empty()){
            result.push_back(minHeap.top().second);
            minHeap.pop();
         }


         return result;
    }
};